Refine target data metadata schema and properties (#33)
reichlab/decisions: decisions/2025-06-17-RFC-target-data-metadata.md

## Summary of `target-data.json` Structure

The `target-data.json` file contains top-level properties that describe expectations across target datasets, with the ability to override these defaults for specific dataset types.

### Configuration Hierarchy

Properties can be set at two levels:

1. **Global (top-level)**: Default values that apply to all target dataset types
2. **Dataset-specific** (`time-series`, `oracle-output`): Override global defaults when needed

When a property is not specified at the dataset level (or is set to `null`), the global value is used.

### Top-Level Properties

* `observable_unit`: An array of column names whose unique value combinations define the minimum observable unit across all target datasets. Must only include the `date_col`, `target_col` (if present), and any other task ID columns. If versioned, unique combinations will also take into account the values in the `as_of` column, but `as_of` is never included in the observable unit itself as it is a versioning column, not a task ID. This property is required.

> * `date_col`: The date column name used across time-series, oracle-output, and model-output datasets. This column stores the date on which observed data actually occurred. Expected to be of type `Date`. This property is required.

Reviewer: ```suggestion
* `date_col`: The date column name used across time-series, oracle-output, and model-output (if present) datasets. This column stores the date on which observed data actually occurred. Expected to be of type `Date`. This property is required.
```

            "items": {
                "type": "string"
            }
        },
        "date_col": {
            "description": "Name of the date column across hub data (time-series, oracle-output and model output). This is the column that stores the date on which observed data actually occured.",
            "type": [
                "string",
                "null"
            ],
            "default": null
        },
        "time-series": {
            "type": "object",
            "properties": {
                "non_task_id_schema": {
                    "type": "object",
                    "uniqueItems": true,
                    "description": "Key-value pairs of non-task ID column names and data types found in time-series data. Include any columns in the time-series data that does not correspond exactly to a task ID. If an `as_of` column is included, it should be specified here as well.",
                    "examples": [
                        {
                            "location_name": "character"
                        },
                        {
                            "date": "Date"
                        }
                    ],

>                     "additionalProperties": {

Reviewer 2: should this be additional_metadata?
```suggestion
                    "additional_metadata": {
```

Reviewer 2: And, maybe this should be moved to the end of the time-series set of properties?

Author: This `additionalProperties` is referring to the contents of the `non_task_id_schema` property which are required to allow properties whose names we don't know a priori (i.e. task ID names)

                    "description": "Names of task IDs whose unique value combinations define an observable unit in oracle output data. Each combination of values must be unique once combined with output type IDs. Can be used to override default observable unit in situations where some output types require additional task ID value to map onto target data.",
                    "type": "array",
                    "uniqueItems": true,
                    "items": {
                        "type": "string"
                    }
                },
                "versioned": {
                    "description": "Indicates whether oracle-output data are versioned using `as_of` dates. If true, the data is expected to have a date `as_of` column that indicates the version of each data point.",
                    "type": "boolean",
                    "default": false
                }
            },
            "additionalProperties": false
        },
        "additional_metadata": {
            "description": "Optional property in which any type of custom metadata can be stored.",
            "type": "object",
            "additionalProperties": true
        }
    },

>     "required": [

Reviewer 2: I might suggest `"versioned"` to be required, just to force people to be explicit about it? (don't feel that strongly about this. if we do this, should apply above to time-series properties as well.)

Author: See overall suggestion below about improving the config usability. Regarding this specific field, have it non-required and defaulting to false is generally considered sensible.

        "schema_version",

>         "observable_unit",

Reviewer 2: shouldn't observable_unit be optional?

Author: We need `observable_unit` to be required because it's the basis for knowing which tasks IDs are part of the target data.

    "properties": {
        "schema_version": {
            "description": "URL to a version of the Modeling Hub schema target-data-schema.json file (see https://github.com/hubverse-org/schemas). Used to declare the schema version a 'target-data.json' file is written for and for config file validation. The URL provided should be the URL to the raw content of the schema file on GitHub.",
            "examples": [
                "https://raw.githubusercontent.com/hubverse-org/schemas/main/v6.0.0/target-data-schema.json"
            ],
            "type": "string",
            "format": "uri"
        },
        "observable_unit": {
            "description": "Names of columns whose unique value combinations define the minimum observable unit across all target type data. Each combination of values must be unique (and in time-series data also unique across `as_of` data versions if applicable). The majority are expected to correspond to task ID names but may include other columns as well (e.g., the `date_col` column).",
            "type": "array",
            "uniqueItems": true,
            "items": {
                "type": "string"
            }
        },
        "date_col": {

>             "description": "Name of the date column across hub data (time-series, oracle-output and ideally model-output). This is the column that stores the date on which observed data actually occurred.",

Reviewer: ```suggestion
            "description": "Name of the date column across hub data (time-series, oracle-output and model-output (if present)). This is the column that stores the date on which observed data actually occurred.",
```
